flow_spec announcement and NDPI packet inspection

dalibort opened the pull request:
Hi,

I have identified and solved bugs related to:
- ExaBGP v4 announcement (the code producing match{} etc code was commented out )
- inspection of captured DDoS attack packets - the file was closed after the first packet was processed. So the attack classification depended only on the first packet in the dump

Note: I am new to github pull requests and C++ to ...

pavel-odintsov commented:
Hello!

That's cool! Have you evaluated code with current exabgp 4.0 release?

On Monday, 8 August 2016, dalibort <notifications@github.com> wrote:

> Hi,
>
> I have identified and solved bugs related to:
>
>    - ExaBGP v4 announcement (the code producing match{} etc code was
>    commented out )
>    - inspection of captured DDoS attack packets - the file was closed
>    after the first packet was processed. So the attack classification depended
>    only on the first packet in the dump
>
> Note: I am new to github pull requests and C++ to ...
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/pavel-odintsov/fastnetmon/pull/540
> Commit Summary
>
>    - - serialize_single_line_exabgp_v4_configuration() now produces
>    announcement compatible
>
> File Changes
>
>    - *M* src/bgp_flow_spec.h
>    <https://github.com/pavel-odintsov/fastnetmon/pull/540/files#diff-0>
>    (6)
>    - *M* src/fastnetmon.cpp
>    <https://github.com/pavel-odintsov/fastnetmon/pull/540/files#diff-1>
>    (46)
>
> Patch Links:
>
>    - https://github.com/pavel-odintsov/fastnetmon/pull/540.patch
>    - https://github.com/pavel-odintsov/fastnetmon/pull/540.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pavel-odintsov/fastnetmon/pull/540>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/ACnfZjPoy2b_U8YQkNjtQTzAybuAa8Jtks5qdzRCgaJpZM4JfEE5>
> .
>


-- 
Sincerely yours, Pavel Odintsov

dalibort commented:
HIo

Yes, I am using v4.0 of ExaBGP. I had no opportunity to check the flow spec announcement in real case (waiting for the proper DDoS)... 
I did some debug printing today and found that the flows_spec messa ge was created improperly - semicolon missing.  So I updated the bgp_flow_spec.h code again. Not sure I have to fill another pull request on GitHub? But it seems this pull requests was automatically updated..

pavel-odintsov commented on src/fastnetmon.cpp:
```diff
@@ -3651,21 +3651,31 @@ void produce_dpi_dump_for_pcap_dump(std::string pcap_file_path, std::stringstrea
 
+    char buff[256];
+                                          
+    snprintf(&buff[0],sizeof(buff)-1,"DPI pkt stats: total:%llu DNS:%llu NTP:%llu SSDP:%llu SNMP:%llu",
```
Could you replace this code by C++ version of sprintf? You could use something like: 
```
logger << log4cpp::Priority::INFO << "DNS " << dns_amplification_packets ....
```

pavel-odintsov commented on src/fastnetmon.cpp:
```diff
@@ -3581,20 +3581,29 @@ void produce_dpi_dump_for_pcap_dump(std::string pcap_file_path, std::stringstrea
     uint64_t ssdp_amplification_packets = 0;
     uint64_t snmp_amplification_packets = 0;
 
+    struct ndpi_id_struct *src = NULL;
```
I'm not sure about correctness of this code because we need to keep state for whole file. Per packet initilization is useless because we lost state about tcp / flows.

dalibort replied:
Hi,
if you are talking about src/dst/flow ndpi vars - your original code allocates and zeroes them per packet too. I am not sure what full tracking is for in this case. I think here we don;t need to analyze tcp and flows.  In this time the attack was already analyzed and detected. This code only tries to match the packets with well known amplification attacks.
Maybe I am missing something?

pavel-odintsov replied:
Yep, it will work per packet basis here but in common case I prefer to keep
one instance per analyzed file.

On Wednesday, 10 August 2016, dalibort <notifications@github.com> wrote:

> In src/fastnetmon.cpp
> <https://github.com/pavel-odintsov/fastnetmon/pull/540#discussion_r74325954>
> :
>
> >      uint64_t dns_amplification_packets = 0;
> >      uint64_t ntp_amplification_packets = 0;
> >      uint64_t ssdp_amplification_packets = 0;
> >      uint64_t snmp_amplification_packets = 0;
> >
> > +    struct ndpi_id_struct *src = NULL;
>
> Hi,
> if you are talking about src/dst/flow ndpi vars - your original code
> allocates and zeroes them per packet too. I am not sure what full tracking
> is for in this case. I think here we don;t need to analyze tcp and flows.
> In this time the attack was already analyzed and detected. This code only
> tries to match the packets with well known amplification attacks.
> Maybe I am missing something?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/pavel-odintsov/fastnetmon/pull/540/files/fcfcc0e6744db22c778cf51341966700c727ffc1#r74325954>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACnfZrLveF6Fkpbwlwt4goYqndb4F8nCks5qejc8gaJpZM4JfEE5>
> .
>


-- 
Sincerely yours, Pavel Odintsov

dalibort commented:
Hi,
I have replaced the snprint() by a C++ code. An I have moved the memset for NDPI flow variable before the packet processing loop. Mainly to prevent memory leaks - it looks like NDPI can do some internal memory allocations which would be lost when flow variable content is zeroed. 
As you can see I allowed to generate multiple flowspec rules for the case that the attack to one IP is multiprotocol one.
Note: The flowspec code requires some additonal changes - like to fire a full IP ban in case the type of attack is not recognised. I am trying to find the proper way how to do it.

pavel-odintsov commented:
Well, my CI is broken. I need some time for fixing it :)

dalibort commented:
Hi,
I have just found that not reinitializing the flow structure inside packet loop is not a good idea. It caused that all 500 packets were recognized as NTP although there is 7 ICMP packet in the dump. Memsetting the flow var before each inspection causes the detection works fine again.

pavel-odintsov commented:
Hello!

Cool! So I fixed Travis CI. Could you commit any fix for triggering Travis build? 

Thanks!

dalibort commented:
Hi,
I commited some changes (commmented out lines removal)

pavel-odintsov commented:
Well, looks great! But I want to divide your changes into 3 different PR. Could you create one PR for each freature below:
- One for changes inside function serialize_single_line_exabgp_v4_configuration
- One for changes related with ndpi
- One for changes related to processing of flow spec announces inside separate function: "launch_bgp_flow_spec_rule"

Could you?

But I'm not sure about changes related with detection logic of mixed attacks. And I want to discuss it later.

I have idea to implement lua api for this part of logic. Then we could move logic outside C++ code.

dalibort commented:
after some git fighting it looks like I had a success to create 3 branches each for the 'isolated' code changes. I fired 3 pull request for them.

Mixed attacks - if you mean simultaneous amplification attacks (my multiflow branch aka launch_bgp_flow_spec_rule). My idea was - it is easy to change the flow_spec filter building code so why don't try it? It is not so important (never seen combined amplification attack yet). 
More important is (at least for me) to have a possibility to filter low speed attack by flow_spec but when the attack is high volume (and upstream links would be overloaded) it should be stopped by RTBH. But I don't know what to do with low volume filterable attack which goes high volume later (probably undetectable because FastNetMon receives no attack traffic due to the filter)

We are an ISP and we observe attacks to our customers which looks like one game players tries to cut another player off the network. These attacks are usually low volume ones and when I ban the IP completely I am helping the attacker to stop his victim from using the internet. If I am able to filter the attack it is better.
For another hand last week we had 30+ high volume attacks (one of our upstreams said 70+gbps and the attack went through multiple upstreams) which were successfully RTBHed by FastNetMon. So if I have to decide whether to use RTBH or flow_spec I must vote for RTBH.
Current problem with flow_spec in FastNetMon is that it looks only me is using/testing it now. There is no code to withdraw the rules, there is no fallback code to RTBH, etc.

PS: I will create another pull request which contains pcap processing made completely in memory (packet_storage.h changes and produce_dpi... changes)

pavel-odintsov commented:
I think this PR is outdated and was divided into few smaller PR's. Closing it.
